app/src/views/features/rules/RuleEditor/components/Header/ActionButtons/actions/index.js

```diff
@@ -9,12 +9,29 @@ import Logger from "lib/logger";
 import * as Sentry from "@sentry/react";
 import { detectUnsettledPromise } from "utils/FunctionUtils";
 import { migrateRuleToMV3 } from "modules/extension/utils";
+import { runMinorFixesOnRule } from "utils/rules/misc";
+import { transformAndValidateRuleFields } from "../CreateRuleButton/actions";
+import { HTML_ERRORS } from "../CreateRuleButton/actions/insertScriptValidators";
+import { actions } from "store";
+import { ToastType } from "componentsV2/CodeEditor/components/EditorToast/types";
+import { toast } from "utils/Toast";
 
-export const saveRule = async (appMode, ruleObject, callback) => {
+export const saveRule = async (appMode, dispatch, ruleObject) => {
   let ruleToSave = cloneDeep(ruleObject);
-
   delete ruleToSave["schemaVersion"];
-  ruleToSave = migrateRuleToMV3(ruleToSave).rule;
+
+  //Pre-validation: regex fix + trim whitespaces
+  const fixedRuleData = runMinorFixesOnRule(dispatch, ruleToSave);
+  //Syntactic Validation
+  const syntaxValidatedRule = await validateSyntaxInRule(dispatch, fixedRuleData);
+
+  if (!syntaxValidatedRule) {
+    return;
+  }
+
+  const parsedRuleData = syntaxValidatedRule || fixedRuleData;
+
+  ruleToSave = migrateRuleToMV3(parsedRuleData).rule;
```

Comment:
moving these checks here is also making them run on template import and status toggle. We already have an unknown performance issue in status toggle which makes it take unexpectedly long times, I would not want to add unnecessary operations to worsen the experience

Reply:
is this absolutely necessary?